In [14]:
# decision trees 
import pandas as pd
import numpy as np

df = pd.read_csv("Student_Performance.csv")
df["Performance Class"] = (df["Performance Index"] >= 60).astype(int)

# Encoding Categorical Variables
# Convert categorical variables (Extra-curricular activities feature) to numeric variables (0/1)
df['Extracurricular Activities'] = df['Extracurricular Activities'].map({'Yes': 1, 'No': 0})

X = df.drop(["Performance Index", "Performance Class"], axis=1)
y = df["Performance Class"]

# Split the data into training and testing sets
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)
from sklearn.tree import DecisionTreeClassifier

model = DecisionTreeClassifier(max_depth=5)
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

from sklearn.metrics import accuracy_score
print("Accuracy:", accuracy_score(y_test, y_pred))

Accuracy: 0.962


In [ ]:
# random forests 
import pandas as pd
import numpy as np

df = pd.read_csv("Student_Performance.csv")
df["Performance Class"] = (df["Performance Index"] >= 60).astype(int)

# Encoding Categorical Variables
# Convert categorical variables (Extra-curricular activities feature) to numeric variables (0/1)
df['Extracurricular Activities'] = df['Extracurricular Activities'].map({'Yes': 1, 'No': 0})

X = df.drop(["Performance Index", "Performance Class"], axis=1)
y = df["Performance Class"]

# Split the data into training and testing sets
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)
from sklearn.ensemble import RandomForestClassifier

model = RandomForestClassifier(n_estimators=100)
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

print("Accuracy:", accuracy_score(y_test, y_pred))

Accuracy: 0.9705


In [20]:
# k-means clustering
import pandas as pd
import numpy as np

df = pd.read_csv("Student_Performance.csv")
df["Performance Class"] = (df["Performance Index"] >= 60).astype(int)

# Encoding Categorical Variables
# Convert categorical variables (Extra-curricular activities feature) to numeric variables (0/1)
df['Extracurricular Activities'] = df['Extracurricular Activities'].map({'Yes': 1, 'No': 0})

X = df.drop(["Performance Index", "Performance Class"], axis=1)
y = df["Performance Class"]

# Split the data into training and testing sets
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)
from sklearn.cluster import KMeans
kmeans = KMeans(n_clusters=2, random_state=42)
kmeans.fit(X)

labels = kmeans.labels_

print("Accuracy:", max(
    accuracy_score(y, labels),
    accuracy_score(y, 1 - labels)
))

Accuracy: 0.8869


In [6]:
# neural networks
import pandas as pd
import numpy as np

df = pd.read_csv("Student_Performance.csv")
df["Performance Class"] = (df["Performance Index"] >= 60).astype(int)

# Encoding Categorical Variables
# Convert categorical variables (Extra-curricular activities feature) to numeric variables (0/1)
df['Extracurricular Activities'] = df['Extracurricular Activities'].map({'Yes': 1, 'No': 0})

X = df.drop(["Performance Index", "Performance Class"], axis=1)
y = df["Performance Class"]

from sklearn.neural_network import MLPClassifier
from sklearn.preprocessing import StandardScaler

# scale
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# train-test split
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, random_state=42)

# model
model = MLPClassifier(hidden_layer_sizes=(10,), max_iter=1000)

model.fit(X_train, y_train)

y_pred = model.predict(X_test)

from sklearn.metrics import accuracy_score
print("Accuracy:", accuracy_score(y_test, y_pred))

Accuracy: 0.9744


Observations

### Decision Tree Observations
* The decision tree model achieved slightly lower accuracy (~95.5%) compared to logistic regression (~97.6%) on the same dataset and split.
* This suggests that the dataset is largely **linearly separable**, where a linear model like logistic regression is sufficient to capture the underlying patterns.
* Decision trees introduce **non-linear decision boundaries** through recursive splitting, which may not be necessary for this dataset and can lead to reduced generalization performance.
* With a constrained depth (`max_depth=3`), the tree remains relatively simple. Increasing depth would likely improve training accuracy but risks **overfitting**, where the model memorizes training data rather than learning general patterns.
* Unlike logistic regression, decision trees do not require feature scaling, as they rely on threshold-based splits rather than distance or gradient-based optimization.
* The structure of the tree (root split) can be used to identify the most important feature, providing interpretability in terms of decision rules.
* Overall, logistic regression outperforms decision trees here, indicating that **model simplicity aligns well with the structure of the data**.
* Increasing the depth of the decision tree from 3 to 5 improved accuracy (from ~95.5% to ~96%), indicating that a slightly more complex model captures additional structure in the data.
* However, further increasing depth caused accuracy to plateau and then decrease, demonstrating the onset of **overfitting**, where the model begins to memorize training data rather than generalizing.
* This behavior illustrates the **bias–variance tradeoff**:
  * Low depth → underfitting (high bias)
  * Optimal depth → good generalization
  * High depth → overfitting (high variance)
* The optimal depth (~5) suggests that the dataset has **moderate complexity**, but does not require highly complex non-linear models.
* Compared to logistic regression, even the optimized decision tree performs slightly worse, reinforcing that the underlying relationships in the dataset are largely linear.

### Random Forest Observations

* Random Forest achieved slightly higher accuracy (~97.05%) compared to a single decision tree (~96%), indicating improved generalization.
* A single decision tree tends to overfit the training data due to high variance. Random Forest mitigates this by training multiple trees on different subsets of data and features.
* By aggregating predictions from multiple trees (majority voting), Random Forest reduces variance and produces more stable and reliable predictions.
* The improvement over decision trees demonstrates how ensemble methods can enhance performance without significantly increasing model complexity from a user perspective.
* However, the performance is still comparable to logistic regression, suggesting that the dataset is largely linearly separable and does not strongly require complex non-linear models.
* Overall, Random Forest provides a balance between flexibility (non-linear modeling) and robustness (reduced overfitting).

### K-Means Clustering Observations

* K-Means is an unsupervised learning algorithm that groups data points based on similarity without using labels.
* It works by iteratively assigning points to the nearest cluster centroid and updating centroids based on the mean of assigned points.
* When applied to this dataset with K=2, the resulting clusters roughly align with the binary performance classes, indicating that the data has a naturally separable structure.
* However, since K-Means does not use labels, cluster assignments are arbitrary and may need alignment (e.g., flipping labels) for comparison.
* The performance is 88.7% ; slightly lower than supervised models, as K-Means does not explicitly learn class boundaries.
* This experiment highlights the difference between supervised and unsupervised learning, where structure is discovered rather than learned from labeled outcomes.